crates/sound/src/stream.rs

```diff
@@ -0,0 +1,305 @@
+///         |              |--------------------->|         |
+///         |              |           |          |         |
+///         |              |---------->|          |         |
+///         |              |           |          |         |
+///         |              |           |          |-------->|
+///         |              |           |          |         |
+///         |              |           |          |<--------|
+///         |              |           |          |         |
+///         |              |           |<-------------------|
+///         |              |           |          |         |
+///         |<-------------------------|          |         |
+///         |              |           |          |         |
+///         |              |<----------|          |         |
+/// ```
+#[derive(Debug, Default, Copy, Clone)]
+pub enum PCMState {
+    #[default]
+    #[doc(alias = "VIRTIO_SND_R_PCM_SET_PARAMS")]
+    SetParameters,
+    #[doc(alias = "VIRTIO_SND_R_PCM_PREPARE")]
+    Prepare,
+    #[doc(alias = "VIRTIO_SND_R_PCM_RELEASE")]
+    Release,
+    #[doc(alias = "VIRTIO_SND_R_PCM_START")]
+    Start,
+    #[doc(alias = "VIRTIO_SND_R_PCM_STOP")]
+    Stop,
+}
+
+macro_rules! set_new_state {
+    ($new_state_fn:ident, $new_state:expr, $($valid_source_states:tt)*) => {
+        pub fn $new_state_fn(&mut self) -> Result<()> {
+            if !matches!(self, $($valid_source_states)*) {
+                return Err(Error::InvalidStateTransition(*self, $new_state));
+            }
+            *self = $new_state;
+            Ok(())
+        }
+    };
+}
+
+impl PCMState {
+    pub fn new() -> Self {
+        Self::SetParameters
+    }
+
+    set_new_state!(
+        set_parameters,
+        Self::SetParameters,
+        Self::SetParameters | Self::Prepare | Self::Release
+    );
+
+    set_new_state!(
+        prepare,
+        Self::Prepare,
+        Self::SetParameters | Self::Prepare | Self::Release
+    );
+
+    set_new_state!(start, Self::Start, Self::Prepare | Self::Stop);
+
+    set_new_state!(stop, Self::Stop, Self::Start);
+
+    set_new_state!(release, Self::Release, Self::Prepare | Self::Stop);
+}
+
+impl std::fmt::Display for PCMState {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        use PCMState::*;
+        match *self {
+            SetParameters => {
+                write!(fmt, "VIRTIO_SND_R_PCM_SET_PARAMS")
+            }
+            Prepare => {
+                write!(fmt, "VIRTIO_SND_R_PCM_PREPARE")
+            }
+            Release => {
+                write!(fmt, "VIRTIO_SND_R_PCM_RELEASE")
+            }
+            Start => {
+                write!(fmt, "VIRTIO_SND_R_PCM_START")
+            }
+            Stop => {
+                write!(fmt, "VIRTIO_SND_R_PCM_STOP")
+            }
+        }
+    }
+}
+
+/// Internal state of a PCM stream of the VIRTIO Sound device.
+#[derive(Debug)]
+pub struct Stream {
```

Comment:
In our approach, we split the stream structure into information that is returned during GET_INFO and information that is used during SET_PARAM. For example, PCMParams is set during SET_INFO so we let the audio-backend handle that. We assume that the audio-backed will accept any configuration. Also, the state of a stream is stored in the audio-backend. We have extended the audio-backend interface to get the state of a stream by the `get_stream_state()` method. The device verifies that a command is executed in the right state. The audio backend decides the next state based on the spec. The device accesses the `stream.state` only for reading whereas the audio backend accesses it for writing and reading. This is because when a cmd is issued in a given state, different audio_backends may decide different transitions as far as they conform to the virtio spec.

Reply:
If you look at alsa.rs, you will see that the audio backend is handling the state transitions here as well.